### PR TITLE
✨ server: implement card statement

### PR DESCRIPTION
**Pull request by nfmelendez**
## Summary by CodeRabbit

* **New Features**
  * Generate and download card statements as PDF
  * Filter activity by maturity date for clearer transaction views
  * Activity responses now include maturity-aware data

* **Bug Fixes**
  * Fixed iterator consumption issue in activity processing

* **Tests**
  * Added tests for maturity filtering and PDF statement rendering

* **Chores**
  * Enabled server-side PDF rendering and JSX, added React linting, workspace catalog entries, and release changesets

---

<a href="https://app.devin.ai/review/exactly/exa/pull/770" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open with Devin
  </picture>
</a>

<h3>Greptile Summary</h3>

implements server-side pdf card statement generation with maturity-based filtering. adds `maturity` query parameter to activity endpoint with content negotiation (json/pdf via accept header). when pdf is requested, renders transactions as statement using react-pdf. filtering logic recalculates installments and amounts for specific maturity dates across both panda and cryptomate card providers.

## key changes
- content negotiation using `hono/accepts` to support `application/pdf` response format
- maturity filtering for `BorrowAtMaturity` events and card transactions
- statement pdf component with purchases and repayment sections
- transaction scoping ensures users only see their own cards on statements
- tests validate pdf rendering, maturity filtering, and cross-user isolation

## areas requiring attention
several previous review threads identified issues that need verification or resolution (see inline comments for details).

<h3>Confidence Score: 3/5</h3>

- safe to merge with moderate risk - complex filtering logic with edge cases noted in previous reviews
- comprehensive test coverage and security measures (transaction isolation) are strong, but complex maturity filtering logic has several edge cases flagged in previous threads that warrant careful verification
- server/api/activity.ts requires careful review of maturity filtering logic and edge cases mentioned in previous review threads

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| server/api/activity.ts | adds pdf statement generation with content negotiation and maturity-based filtering; complex logic for filtering card transactions and installments by maturity date |
| server/utils/Statement.tsx | new react-pdf component for rendering card statements; clean implementation with proper style definitions |
| src/utils/server.ts | adds overloaded `getActivity` function to handle pdf responses; properly typed for both json and pdf return types |
| server/test/api/activity.test.ts | comprehensive tests for maturity filtering and pdf generation; includes security test for transaction isolation across users |

</details>

<details><summary><h3>Sequence Diagram</h3></summary>

```mermaid
sequenceDiagram
    participant Client as Mobile App
    participant API as activity API
    participant DB as Database
    participant Chain as Blockchain
    participant PDF as React-PDF

    Client->>API: GET /activity?maturity=X&accept=application/pdf
    API->>API: Parse accept header with hono/accepts
    API->>DB: Query credentials, cards, transactions
    API->>Chain: Fetch BorrowAtMaturity events
    API->>Chain: Fetch block timestamps
    API->>API: Filter transactions by maturity
    API->>API: Build statement data (purchases + repayments)
    API->>PDF: renderToBuffer(Statement component)
    PDF-->>API: PDF binary (Uint8Array)
    API-->>Client: Return PDF with content-type: application/pdf
    
    Client->>API: GET /activity?maturity=X (JSON)
    API->>API: Parse accept header (defaults to JSON)
    API->>DB: Query credentials, cards, transactions
    API->>Chain: Fetch events and timestamps
    API->>API: Filter by maturity, recalculate installments
    API-->>Client: Return filtered JSON array
```
</details>

<sub>Last reviewed commit: 2d6a53a</sub>

### CODE REVIEW

File: server/utils/Statement.tsx [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+        positionAmount: number;
+        timestamp: string;
+      }
+    | { description: string; installments: { amount: number; current: number; total: number }[]; timestamp: string }
+  )[];
+  lastFour: string;
+  maturity: number;
+}) => {
+  const dueDate = new Date(statement.maturity * 1000).toLocaleDateString("en-CA");
+  const statementDate = new Date((statement.maturity - MATURITY_INTERVAL) * 1000).toLocaleDateString("en-CA");
```
[the remark this answers is not in the document]
Comment (nfmelendez) [reply server/utils/Statement.tsx]: could you explain more what you mean by " locale-specific behavior "

File: server/api/activity.ts [hunk 2]
```diff
@@ -63,11 +66,14 @@ export default new Hono().get(
   auth(),
   vValidator(
     "query",
-    optional(object({ include: optional(union([ActivityTypes, array(ActivityTypes)])) }), {}),
+    optional(
+      object({ include: optional(union([ActivityTypes, array(ActivityTypes)])), maturity: optional(string()) }),
+      {},
+    ),
```
[the remark this answers is not in the document]
Comment (nfmelendez) [reply server/api/activity.ts]: exist in valibot a function similar to regex(/^\d+$/) ?

File: server/api/activity.ts [hunk 3]
```diff
@@ -245,54 +255,133 @@ export default new Hono().get(
...
+        const timestamp = timestamps.get(blockNumber);
+        if (timestamp) return { ...event, timestamp: new Date(Number(timestamp) * 1000).toISOString() };
+        captureException(new Error("block not found"), {
+          level: "error",
+          contexts: { event: { ...event, timestamp } },
+        });
+      }),
+    ]
+      .filter(<T>(value: T | undefined): value is T => value !== undefined)
+      .toSorted((a, b) => b.timestamp.localeCompare(a.timestamp) || b.id.localeCompare(a.id));
```
[the remark this answers is not in the document]
Comment (aguxez) [reply server/api/activity.ts]: We'll skip this change. This creates unnecessary types definitions we need to satisfy the change to use 'flapMap' in this instance.
